File: src/server/importer/importers/n26-importer.ts
import { parse } from 'csv-parse';
import fs from 'node:fs/promises';

import { BankTransactionBase } from '../../bank-transaction/models/bank-transaction.model';
import { Exception } from '../../core/errors/exception';
import { TransformHelper } from '../transform.helper';
import { ParserFactory } from './importer.service';

export class N26Importer implements ParserFactory {
    mapper = new TransformHelper<number>(
        new Map([
            ['movementName', 1],
            ['date', 0],
            ['dateValue', 0],
            ['value', 5],
        ]),
    );

    async *create(filePath: string): AsyncGenerator<BankTransactionBase> {
        try {
            const fileData = await fs.readFile(filePath);
            const csvData = parse(fileData, { autoParseDate: true, delimiter: ',', fromLine: 2 });

            for await (const row of csvData) {
                yield this.mapper.map(row as Record<number, unknown>);
            }
        } catch (error) {
            if (typeof error === 'object' && error && 'code' in error && error.code === 'ENOENT') {
                throw new Exception('E10006', 'N26 file not found', { filePath }, error);
            }
            throw error;
        }
    }

    fileRegex = /n26-csv-transactions.*\.csv/u;

    key = 'n26/es';
}

File: swift-server/app/importers/importers/n26.importer.swift
import CSV
import Foundation
import Vapor

class N26Importer: ParserFactory {
    let transformHelper: TransformHelper<[String]>
	let key: String = "n26/es"
	let fileRegex: String = "n26-csv-transactions.*\\.csv"

	init() {
		let fieldsMap = FieldsMap<Int>(
			movementName: 1,
			date: 0,
			dateValue: 0,
			value: 5
		)
		self.transformHelper = TransformHelper(fieldsMap, dateFormat: "yyyy-MM-dd")
	}

	func create(filePath: String) -> AsyncThrowingStream<PartialBankTransaction, Error> {
		AsyncThrowingStream { continuation in
			Task {
				do {
					let csv = try parseCsv(filePath: filePath)

					var lineCounter = 0
					while let row = csv.next() {
						lineCounter += 1
						do {
							let transaction = try self.transformHelper
								.map(row)
							continuation.yield(transaction)
						} catch {
							throw Exception(
								.E10009,
								context: ["line": lineCounter],
								cause: error)
						}
					}

					continuation.finish()
				} catch {
					continuation.finish(throwing: error)
				}
			}
		}
	}
}
